API changes to make certain things feel more natural in Zig

bronter opened the pull request:
Related to the discussion in #23:

- [ ] use slices instead of separate count and pointer
- [ ] return error instead of null when null actually indicates a failure
- [ ] return the value instead of putting it into a pointer in some functions like surface.getCurrentTexture
- [ ] use a zig bool instead of WGPUBool
- [ ] pass the descriptor structs, and the structs they may contain, directly as value instead of as pointers
- [ ] u8 slices instead of StringViews
- [ ] Return error unions rather than `wgpu.Status`
- [x] Get rid of `Procs` structs since they are unused and difficult to maintain
- [x] Use packed structs to represent `WGPUFlags`

bronter committed:
Initial work on making Instance structs/methods more idiomatic

bronter commented:
I just want it to be be somewhat obvious which method generated an error, so that if I have an application that has a `fn render(ctx: *RenderContext) !void {}, and `context.render()` returns an error I know which method call within `render()` generated the error before I even look at the stack trace.

TotoShampoin committed:
Similar initial work for Device structs/methods

bronter commented:
Yeah, I should've run tests on that, will update it to use `[]const u8`.

bronter committed:
Merge pull request #25 from TotoShampoin/idiomatic-zig

Similar initial work as Instance structs/methods, but for Device

bronter committed:
use const for u8 slices

TotoShampoin committed:
More idiomatic Adapter structs and methods

TotoShampoin committed:
Fixed typo

bronter committed:
Merge pull request #26 from TotoShampoin/idiomatic-zig

More idiomatic Adapter structs and methods

TotoShampoin commented:
I believe that the 2nd userdata is meant to be used by engines that use webgpu anyway, so yeah

Another thing you could do, if you really want to keep 2 userdatas (as this is still a webgpu interface), is put them in a tuple of opaque pointers, and pass that as parameter 1

TotoShampoin commented:
Side note, `anyopaque` isn't confusing if you think that it's equivalent to C's `void*`, which means "pointer to an unknown type"

bronter committed:
Try new callback format on Instance.requestAdapter()

bronter committed:
Implement new callback format for Adapter.requestDevice()

bronter committed:
Implement toWGPU() method for structs in instance.zig

TotoShampoin committed:
Implement toWGPU() method for structs in adapter.zig and device.zig

TotoShampoin committed:
Implement toWGPU() method for structs in adapter.zig

TotoShampoin committed:
Implement toWGPU() method for structs in device.zig

TotoShampoin committed:
Merge branch 'idiomatic-zig' of https://github.com/TotoShampoin/wgpu_native_zig into idiomatic-zig

bronter committed:
Merge pull request #28 from TotoShampoin/idiomatic-zig

Implement toWGPU() method for structs in adapter.zig device.zig

TotoShampoin commented:
Another con I see to doing this is we stray further away from the native webgpu api, which, if I wanted that, I would have just gone with the Zig-Gamedev implementation. 

Using .toWGPU was my initial idea to do some separation of concerns, but if we keep in mind that the motive here is to convert idiomatic zig structs to what WebGPU expects, then I have a few ideas:

We can create a slice using just the pointer and the length and inversely, as long as the original variables are not deallocated (`const slice = your_buffer[0..length]`). So one thing that *could* be done is simply to not guarantee that the values will be preserved. 

Maybe there would be a .toAllocatedWGPU and a .toUnsafeWGPU, where the latter does exactly that. That way, there would be an implementation for both preferences (which seems to be in line with what Zig's std tends to do), and the interface could just use .toUnsafeWGPU internally.  
Like, I don't think that WebGPU will face dangling pointers after the bind group will be created as I believe it does not need the descriptor once that's done? I could be wrong though. Or maybe I misunderstood where you're having allocation issues.

bronter commented:
I'd agree with not straying to far from the native webgpu api.

The issue is with slices of wrapper structs. If I have an array of `BindGroupEntry`, and `BindGroupEntry` has a different size or memory layout than `WGPUBindGroupEntry`, then I can't trivially convert a slice of `BindGroupEntry` to a slice of `WGPUBindGroupEntry`; I'd have to allocate memory and construct a new array of `WGPUBindGroupEntry`.

bronter commented:
So as far as I can tell, we either have to avoid slices of wrapper structs with different sizes than the native structs, or pass in an allocator.

TotoShampoin commented:
Oohh, now I understand the issue...

Well, not only would you have to avoid different sizes than the native structs, but you'd also have to avoid different layouts. As in, if a zig slice stores the pointer first but wgpu-native stores the length first, we're cooked.

Another issue is that this is extra work at runtime for allocating and copying or converting the same data...

bronter committed:
Merge branch 'main' into idiomatic-zig

bronter commented:
Another option would be to use a pattern similar to what we have for shader module descriptors in `shader.zig`. Instead of making `BindGroupEntry` into a wrapper struct with a `toWGPU()` method, we could keep `BindGroupEntry` as the native struct, but provide a function that can construct it taking a wrapper struct (like `WrappedBindGroupEntry`), for example:
```zig
fn bindGroupEntry(entry: WrappedBindGroupEntry) BindGroupEntry {
  //...
}
```
Then BindGroupDescriptor can take a slice of `BindGroupEntry`, and it stills looks fairly normal to construct, like:
```zig
const bind_group = device.createBindGroup(&BindGroupDescriptor {
  //...other stuff in the descriptor
  entries: &[_]BindGroupEntry {
    bindGroupEntry(.{
      //...
    }),
    bindGroupEntry(.{
      //...
    }),
  },
});
```

bronter commented:
I could make the function `inline`, though if the structs are all on the stack I'd be surprised if it wasn't optimized out anyway. I'll see if there's a way to test that.

TotoShampoin commented:
Well, it's something I found out by using a debugger, finding out that a lot of unnecessary jumps (especially switches) were dissipated even in debug mode

I don't remember the intricate details, and I have yet to figure out how to debug a zig executable in vscode on windows

bronter committed:
Test and related bugfixes for DeviceDescriptor.toWGPU()

bronter committed:
Run unit tests from root to avoid duplicate test runs

bronter committed:
Merge branch 'main' into idiomatic-zig

bronter committed:
refAllDecls from root

bronter committed:
Merge branch 'main' into idiomatic-zig

TotoShampoin committed:
Using packed structs instead of ORed consts

bronter commented:
Yeah, the padding thing is a bit annoying, though I think it makes sense since Zig tries not to have any hidden behavior and you'd want to be able to see what the rest of the bits are explicitly set to.

I think you're right about all the flag bits being adjacent. Some flags have been added or removed over time, but when that happens the rest of the flags are changed so that there aren't any gaps. The enums sometimes have gaps though, but for those it doesn't really matter.

TotoShampoin committed:
Added a missing `all` const

TotoShampoin committed:
Removed commented out code

bronter committed:
Merge pull request #31 from TotoShampoin/idiomatic-zig

Using packed structs instead of ORed consts

bronter committed:
Remove all Procs

bronter committed:
Added some constructor functions for DeviceLostCallbackInfo and UncapturedErrorCallbackInfo

bronter committed:
Add tests for deviceLostCallbackInfo() and uncapturedErrorCallbackInfo() and fix related bugs

bronter committed:
Tweak CallbackInfo struct construction a bit more

Add init() method so that we don't have both a function and a struct with the same name
